ECS: Rendering, Economy, ++

IllustrisJack opened the pull request:
This is a collapsed branch of `feat/economy`and `dga/integrate-ecs-with-legacy-entity`

## copied from `dga/integrate-ecs-with-legacy-entity` https://github.com/Limit-Theory-Redux/ltheory/pull/355
The goal of this PR is to integrate the ECS with the existing "legacy" entity stack.

The legacy Entity object will essentially be a wrapper over the ECS Entity, where the existing logic can tie in to the ECS via those wrapper functions that already exist (like `self:addRigidBody` etc).

## copied from `feat/economy` https://github.com/Limit-Theory-Redux/ltheory/pull/357
This PR draft serves as a tracker until the implementation is complete.

TODO:
- [x] Create an ECS Entity for every Legacy Entity.
- [x] Integrate the physics RigidBody with the ECS.
- [x] Integrate the mesh/renderable with the ECS.
- [ ] Port all custom renderables to the ECS RenderComponent.
- [x] Introduce a TransformComponent, which is being updated by a PhysicsSystem (that copies RigidBody > Transform)
- [x] Wire up the RenderComponent with an ECS system, rather than being driven by the GameView.

## Other Todo 
- [x] CameraSystem
- [x] PlanetTest reintroduced

dgavedissian committed:
Update RenderComponent to allow multiple meshes with different
materials.

dgavedissian committed:
Replace Entity:addVisibleMesh with a RenderComponent.

Render these components from within GameView:drawScene.

dgavedissian committed:
Fix PhysicsTest and ShipTest

dgavedissian committed:
Integrate the ECS with Legacy.Entity, use the ECS to store the name and rigid body.

IllustrisJack committed:
fix: deep clone

dgavedissian committed:
Scale IsoSphere mesh for atmosphere at gen time.

This avoids us needed to scale through the world matrix.

IllustrisJack committed:
refactor(inventory): use fixed clone method, inventory manager

IllustrisJack committed:
refactor: cleanup trailing and unnecessary comment

IllustrisJack committed:
feat(tags): add tags

dgavedissian committed:
Use RigidBodyComponent instead of the legacy entity when rendering ECS
entities.

IllustrisJack committed:
refactor(economy): remove helper function step & debug infinite trading loop

- this or store *only* the id?

MotorcycleEnjoyer committed:
feat(economy): implement basic trade process

dgavedissian committed:
Move planet properties into a component, rename setState > updateState.

dgavedissian committed:
Clean up some of the abstractions in RenderCoreSystem

IllustrisJack committed:
fix: render test

- this partially fixes the render test, though it´s really laggy

IllustrisJack committed:
chore: add comment for temp code

IllustrisJack committed:
fix: use id of item type

dgavedissian committed:
Fix PlanetTest

dgavedissian committed:
Port Planet to RenderComponent.

This commit also adds a bunch of hacks to the Legacy Material (which I
want to later unify with the new Material)

IllustrisJack committed:
feat(economy): example tag usage for further expanding

IllustrisJack committed:
refactor(tests): marketplace test

IllustrisJack committed:
feat(math): add some quat funcitonality

IllustrisJack committed:
feat(bit): some more bit additions

IllustrisJack committed:
feat(rendering): rendering test via ECS RenderCoreSystem

- still need to decide how to handle method overwrites from Application -> State, maybe we just disallow them

IllustrisJack committed:
feat(gen): ruleset testing, add more standard rulesets

- fully add missing component content
- plus many more fixes and adjustments, too many for one commit (sorry)

IllustrisJack committed:
feat(economy): start using delta timer class

IllustrisJack committed:
refactor(gen): small naming fixes

IllustrisJack committed:
feat(gen): expand rules, entity placement, introduction of gameunit, universe scale

Major overhaul of StandardSolarSystem ruleset: replaced conditional ranges with combined criteria, added 3D orbital parameters, and restructured asteroid ring/belt logic. Introduced UniverseScaleConfig for unit conversions and scaling. Added PositionCalculator for 3D orbital mechanics. Updated enums for new rule types and stability values. Refactored RuleEvaluator to support combined conditions and new rule types. Updated test and type definitions for consistency with new aspect names.

IllustrisJack committed:
feat(rendering): more comprehensive test via ECS Rendering

- uses ECS rendering <3

IllustrisJack committed:
feat(economy): use adjusted delta timer

IllustrisJack committed:
feat(ecs): add initial universe manager prototype

- Scenarios + Rules for generalized generation via manager
- Next step: AI Tasking

IllustrisJack committed:
refactor(gen): more accurate debug prints for unmet conditions

IllustrisJack committed:
feat: set to standard test

IllustrisJack committed:
fix(tests): fix gen, trading tests with new app class struct

IllustrisJack committed:
tweak(economy): tweak pull recovery rate

for easier testing

IllustrisJack committed:
feat(rendering): add axis, rendering (non worky)

- this is due to draw still using Vec3f, needs to support new Position type or Position type needs conversions etc

IllustrisJack committed:
feat(economy): many fixes and improvements (test & system)

IllustrisJack committed:
feat(gen): add initial rulebased solar system gen + test

IllustrisJack committed:
feat: add deltatimer class

IllustrisJack committed:
feat(dtimer): extend delta timer

IllustrisJack committed:
fix(scenarios): remove fixed seed in singe solar system

- we don´t want a fixed seed in a procgen scenario

IllustrisJack committed:
feat(economy): better test charts

IllustrisJack committed:
feat(ecs): itemType cmp -> item cmp & some cleanup

IllustrisJack committed:
refactor(events): some event fixes and docs

IllustrisJack committed:
feat(gen): use enums for everything

IllustrisJack committed:
feat(economy): extensive marketplace testing, system expansion

IllustrisJack committed:
feat(tests): move outdated tests to seperate folder

IllustrisJack committed:
feat(rigidbody): add ext for getPos, getRot

IllustrisJack committed:
feat(ecs): add planet test, allow render fn for rendercmp

- still missing planet materials

IllustrisJack committed:
feat(gen): functional planet rendering test, moved legacy materials

IllustrisJack committed:
feat(gen): planet test mat cloning and alternating textures

IllustrisJack committed:
feat(gen): change test to 1024 texture

IllustrisJack committed:
feat(rendercore): add material ecs data caching, add ring

IllustrisJack committed:
feat(gen): tweak planet shader, clouds etc

- CloudMotionComponent dummy

IllustrisJack committed:
feat(shaders): better clouds

IllustrisJack committed:
feat(planet_test): add better camera options

IllustrisJack committed:
feat(tests): planet test - sizes, cam improvements, registry destroy entity adjustment

IllustrisJack committed:
feat(ecs): camera system additions

IllustrisJack committed:
feat(quat): reintroduce legacy quat method (from look up)

IllustrisJack committed:
feat(legacy): legacy planet, use empty cmp

IllustrisJack committed:
feat(tests): planet test - add orbiting moons

IllustrisJack committed:
feat(shaders): add missing ring debug shader

IllustrisJack committed:
refactor(shaders): better planetring debug shader

IllustrisJack committed:
refactor(shader): initial planetring shader & motion

HaronK commented on engine/lib/phx/src/math/quat.rs:
```diff
@@ -196,38 +244,71 @@ impl Quat {
-        e |= validate_f64(self.0.w as f64);
-        e
+    #[bind(name = "FromAxisAngle")]
+    pub fn from_axis_angle(axis: &Vec3, radians: f32) -> Self {
```
Same here

HaronK committed:
Consolidate ShaderVar classes

HaronK committed:
Support literal and function for value representation. Improve logging

HaronK committed:
Use table key as a texture and shader var name in material definitions. Rename fields

HaronK committed:
Rename AutoShaderVar to DynamicShaderVar

HaronK committed:
Dynamically generate DynamicShaderVar:getValue method

HaronK committed:
Allow table shader values

HaronK committed:
Merge pull request #360 from Limit-Theory-Redux/feat/shader-var

Refactor material definition

IllustrisJack committed:
fix(material): textures not applied/created correctly

HaronK committed:
fix(material): apply textures properly

IllustrisJack committed:
feat: transform updating from rb

IllustrisJack committed:
feat: use updated transform

IllustrisJack committed:
feat: add simple moon shading

IllustrisJack committed:
feat: click raycasting & camera entity focus

IllustrisJack committed:
feat(ecs): stuff

IllustrisJack committed:
feat(engine): fixes and improvements to Vec3f, Quat, Cursor

IllustrisJack committed:
feat(ecs): add camera module (camera controller & manager, freecam)

IllustrisJack committed:
chore: update macos runner

IllustrisJack committed:
chore: update runners in ci.yaml aswell

funky-tm committed:
actionbinding and input manager scaffold

IllustrisJack committed:
feat(ships): add spawnable ecs spaceship

IllustrisJack committed:
feat(cameras): add fp & orbiting camera, camera system

IllustrisJack committed:
refactor: remove unnecessary print

IllustrisJack committed:
feat(cameras): add strategic camera

IllustrisJack committed:
refactor(cameras): rts camera tweaks and improvements

IllustrisJack committed:
fix: clippy error bit rot

IllustrisJack committed:
fix: planet test, misuse of eye in camera manager

funky-tm committed:
added input test

funky-tm committed:
Merge branch 'feat/ecs' into feat/input

IllustrisJack committed:
feat(camera_test): allow changing fov

IllustrisJack committed:
feat: cinematic camera + rendering test fixup

funky-tm committed:
Merge branch 'feat/ecs' into feat/input

funky-tm committed:
replacing camera init

funky-tm committed:
cleanup

IllustrisJack committed:
feat(input): fix gamepad axis states, add delta method, other fixes

IllustrisJack committed:
feat(gc): add experimental manual GC

IllustrisJack committed:
feat: add gc debug info to camera test & planet test

IllustrisJack committed:
fix: cursor grabbing

IllustrisJack committed:
feat: update controllers for cursor grabbing

IllustrisJack committed:
feat: make ship components more generic

IllustrisJack committed:
feat(input: fix gamepad analog values

IllustrisJack committed:
feat(cameras): much better rts camera

IllustrisJack committed:
feat(tests): revive station test, allow station generation

IllustrisJack committed:
fix(rendering): fix textures not rendering

funky-tm committed:
swapping actionbinding types to enums, adding cursor to test, adding actionbinding profiling to test

IllustrisJack committed:
feat(moon): add moon test + nicer moon

IllustrisJack committed:
feat(tests): revive ship test

IllustrisJack committed:
feat: generators folder & correct module init

IllustrisJack committed:
feat(rendering): post processing

- sharpen
- fxaa
- tonemap
- colorgrade
- dither
- radialblur
- config

no warranty included

IllustrisJack committed:
refactor: reset supersample to default woops

IllustrisJack committed:
refactor: optimize color grade shader

IllustrisJack committed:
tweak(postfx): tweak config a little

IllustrisJack committed:
feat(rendering): move fps & frametime tracking to render core

IllustrisJack committed:
fix(winit): borderless fullscreen

IllustrisJack committed:
tweak(renderconfig): change suggested ss range

- anthing above 2x will produce viewport issues, might have to fix in render pipeline

funky-tm committed:
add gamepad cursor action binding, Control.Single enforce, test cleanup

funky-tm committed:
add action binding tap and hold tracking + docking sim to test

HaronK commented on engine/lib/phx/src/engine/engine.rs:
```diff
@@ -145,78 +146,120 @@ impl Engine {
                 WindowMode::BorderlessFullscreen => {
-                    Some(winit::window::Fullscreen::Borderless(None))
+                    // Choose the monitor explicitly (primary monitor recommended)
+                    let monitor = self.winit_window.window().primary_monitor();
```
If possible, it would be better to detect 'active' monitor where game window is now and use it.
Would be also nice to make it configurable and maybe save user choice when window is dragged to another monitor.
Though there should be another PR for this.

IllustrisJack commented on engine/lib/phx/src/engine/engine.rs:
```diff
@@ -145,78 +146,120 @@ impl Engine {
+                        .winit_window
+                        .window()
+                        .current_monitor()
+                        .unwrap_or_else(|| self.winit_window.window().primary_monitor().unwrap());
```
Gotcha. Used unwrap here cause this should really never be the case but we maybe should create a small exception window library (via winit) that we could use here. But also yes, being able to choose would be the best solution.

IllustrisJack committed:
Merge feat/ecs into feat/input to create playable-solarsystem base

IllustrisJack committed:
feat: engine fixes, rendering pipeline, shaders

- Reorder frame stages (Input before Sim before Render) to eliminate camera lag
- Fix mouse scroll/delta persisting between frames after stage reorder
- Extend far plane to 1e8 for distant object visibility
- Add Vec3f scalar arithmetic and RigidBody velocity FFI wrappers
- Add deferred directional lighting pass to render core
- Add star shader (FBM with domain warping, limb darkening, type tinting)
- Add lens flare shader (anamorphic streaks, chromatic ghosts)
- Add travel drive visual effect shaders
- Add 3D map shaders (solid color, orbital trails, hologram vertex)
- Add sun texture for star rendering

IllustrisJack committed:
feat: playable solar system with maps, visuals, and reusable systems

- Add SolarSystemPlayable game state with ECS-generated solar system
- Add SolarSystemVisualizer for procedural planet/moon/star materialization
- Add OrbitalSystem for Kepler-style orbital simulation
- Add 2D system map with zoom/pan, click-to-select, orbit rings
- Add 3D holographic system map with orbital trails and focus tracking
- Add CoordinateRebaser for f64 precision (local origin rebasing)
- Add CelestialLightingSystem, LensFlareSystem, GameplayHUDSystem,
  WorldLabelRenderSystem as reusable extracted systems
- Add TidalLockComponent for close-orbit planets
- Add chase camera controller with ship-relative smooth follow
- Update camera modes, rulesets, and material definitions

IllustrisJack committed:
feat: ship flight, travel drive, gravity wells, autopilot

- Add TravelDriveSystem with charging/active/decelerating states
- Add GravityWellSystem with zone detection and velocity matching
- Add AutoPilotSystem with velocity-target navigation and orbit matching
- Add ShipFlightSystem with FPS indirect aiming and chase mode
- Add PlayerController for camera modes and input routing
- Add ECS components (TravelDrive, AutoPilot, ShipFlightControl) for
  proper per-entity state instead of singleton systems
- Add ship action bindings (thrust, roll, boost, travel drive toggle)
- Zone-based travel drive speed limiting with smooth distance falloff

IllustrisJack committed:
feat: game config for ship physics, travel drive, and camera

- Add formula-based ship flight config (thrust derived from desired speed/drag)
- Add ship hull classes with per-class mass, drag, and maneuverability
- Add travel drive config with zone-based speed limits
- Add autopilot and chase camera config
- Add solar system playable defaults (spawn, textures, orbit camera)
- Add distance/speed formatting helpers to UniverseScaleConfig

IllustrisJack committed:
feat: shader hot-reload system

- Store source paths in Shader for reloading from disk
- Add Shader::reload() with graceful error handling (keeps old shader on failure)
- Add non-panicking try_create_gl_shader/try_create_gl_program for safe recompilation
- Add Cache.ReloadShaders() to reload all cached shaders
- Add Material.ReloadAll() with weak-ref registry to refresh all live materials
- Press F5 in-game to hot-reload all shaders without restarting

IllustrisJack committed:
refactor: centralized input via ActionBindings and CursorManager

- Add CursorManager for unified cursor visibility and grab mode control
- Add GeneralActions (camera cycle, map toggle, autopilot, shader reload)
- Add MapActions (select, pan, drag, zoom, move, rotate) for 2D/3D maps
- Add CameraActions bindings (roll, fast/slow move, mouse capture)
- Migrate FreeCameraController and OrbitCameraController to ActionBindings
- Migrate SystemMap and SystemMap3D to MapActions
- Migrate PlayerController to CursorManager and GeneralActions
- Migrate SolarSystemPlayable input to GeneralActions
- Wire F5 shader hot-reload through ActionBinding system

IllustrisJack committed:
feat: asteroid belts with chunk-based entity spawning

- Add AsteroidBeltRenderer for batch rendering 100k+ asteroids with LOD
- Add AsteroidFieldSystem with Minecraft-style chunk spawning: nearby
  asteroids become real entities (physics, selectable, navigatable),
  distant ones are batch-rendered visuals
- Add AsteroidBeltComponent storing asteroid data, orbit radius, width,
  shared LodMesh, and angular buckets for fast spatial lookup
- Fix asteroid belt position (absolute coords like planets)
- Fix AsteroidMesh.lua require path and GC-managed object cleanup
- Add annulus shader for solid belt ring visualization on 2D map
- Add asteroid dot rendering on map at high zoom with click-to-navigate
- Add belt ring click detection for autopilot to any point on the belt
- Add autopilot ETA display and distance-based drive deceleration
- Add Physics::draw_wireframes_in_range with distance culling (green)
- Add nav target marker and dashed line on 2D map
- Map follows clicked nav position, not belt entity center

IllustrisJack committed:
feat: asteroid rings, mesh pool, map optimizations, intercept autopilot

Asteroid rings:
- Add planetary asteroid rings with same logic as belts
- Rings follow parent planet position each frame
- Ring orbit radius from planet scale (2-5x), no Orbit component
- Fix ring/belt positioning (rings centered on planet, belts at origin)
- Fix universe generator ring position (absolute coords for rebase)

Performance:
- AsteroidMeshPool: 8 cached SDF mesh variants, loaded from disk
- Add Mesh.save() to Rust for mesh caching to res/mesh/asteroid/
- Lazy rotation cache (compute on first render, not at startup)
- Max 200 drawn asteroids per frame per belt/ring
- Render distance proportional to ring size
- Map dot rendering via single-draw-call GPU mesh (mappoints shader)
- Cached dot mesh per belt/ring (never rebuilt)
- Skip walking belt/ring children in map entity collection
- Viewport culling + size cap for annulus/orbit ring quads
- Replace solidcolor3d with solidcolor shader

Map improvements:
- Annulus rings centered on parent (planet for rings, star for belts)
- Separate dot threshold for rings (20px) vs belts (2x screen)
- Nav target marker on actual target, separate intercept marker (cyan)
- White flight path line with distance + ETA label at midpoint
- Asteroid click detection via angular buckets

Autopilot:
- Intercept course via proportional navigation (position delta velocity)
- Cruise speed prediction for stable initial intercept
- Skip leading inside gravity well (velocity matching handles it)
- Lead blend: full intercept far, direct approach close
- Arrival check against actual target, not intercept point
- Stable drive engage/disengage thresholds (no loop)
- Autopilot runs during map (not blocked by input)
- Nav data stored on AutoPilotComponent (interceptPos, distance, eta)
- HUD reads from component (no duplicate computation)

IllustrisJack committed:
fix: asteroid field architecture, realistic scales, autopilot improvements

Architecture:
- OrbitalSystem tracks followers (rings) alongside orbiters
- Rings auto-follow parent planet position every frame
- AsteroidFieldSystem split: updatePositions() every frame, update() rate-limited
- Fixed spawn distance check (was missing belt entity offset for rings)
- Simplified spawned entity: mesh+material, sphere collider, no custom renderFn
- Remove manual ring position update from SolarSystemPlayable

Autopilot:
- Intercept prediction uses expected cruise speed, not current speed
- Skip leading + orbit matching inside gravity wells (fly direct)
- Distance-based close approach (within 5000 units: fly direct)
- Smart speed curve: distance-proportional drive multiplier cap
- Parent-based target tracking for ring asteroids
- Store parentEntity + localOffset on AutoPilot component
- Stable drive engage/disengage thresholds

Scales (globalScale 0.0001):
- Ship: 50m fighter (scale 0.005, mass 5000kg)
- Belt asteroids: 1-50km (scale 0.1-5.0)
- Ring asteroids: 0.5-20km (scale 0.05-2.0)
- Station: 3km (scale 0.3)
- Torque tuned empirically (50), angular drag 20
- Camera distances scaled for 50m ship

Map:
- Clean nav display: TARGET marker (yellow), INTERCEPT marker (cyan)
- Spawned asteroid entities visible as map labels (nearest 10)
- Removed redundant selection line when autopilot active

IllustrisJack committed:
style: fix rustfmt formatting in shader.rs

IllustrisJack committed:
fix: resolve clippy unnecessary_unwrap warnings in luajit-ffi-gen
